ghcid-ng/src/ghci/parse/module_set.rs

```diff
@@ -0,0 +1,81 @@
+use std::collections::HashSet;
+
+use camino::Utf8Path;
+use camino::Utf8PathBuf;
+use miette::Context;
+use miette::IntoDiagnostic;
+
+use super::Module;
+
+/// A collection of source paths, retaining information about loaded modules in a `ghci`
+/// session.
+#[derive(Debug, Clone, Default, PartialEq, Eq)]
+pub struct ModuleSet {
+    map: HashSet<Utf8PathBuf>,
+}
+
+impl ModuleSet {
+    /// Parse a `ModuleSet` from a set of lines read from `ghci` stdout.
+    pub fn from_lines(lines: &str) -> miette::Result<Self> {
+        Ok(Self {
+            map: lines
+                .lines()
+                .map(|line| {
+                    line.parse::<Module>()
+                        .wrap_err("Failed to parse `:show modules` line")
+                        .and_then(|module| canonicalize(&module.path))
+                })
+                .collect::<Result<_, _>>()?,
+        })
+    }
+
+    /// Get the number of modules in this set.
+    pub fn len(&self) -> usize {
+        self.map.len()
+    }
+
+    /// Determine if this set is empty.
+    #[allow(dead_code)]
+    pub fn is_empty(&self) -> bool {
+        self.len() == 0
+    }
+
+    /// Remove all entries from this set, leaving it empty.
+    #[allow(dead_code)]
+    pub fn clear(&mut self) {
+        self.map.clear();
+    }
```

Comment:
i feel like it would be easier to make the `map` field public. i don't think much is gained from information hiding here and then you wouldn't need these methods

Reply:
Not just information hiding, we don't want users to insert things into the set without knowing about it -- in particular, we want all of our paths to be nicely canonicalized.

Reply:
ah okay, never mind then

Reply:
I should eventually write a `CanonicalizedUtf8PathBuf` type, though.